verifier/src/lib.rs

```diff
@@ -82,92 +62,98 @@ pub fn verify(
     Ok(security_level)
 }
 
-/// Identical to [`verify`], with additional verification of any precompile requests made during the
-/// VM execution. The resulting aggregated precompile commitment is returned, which can be compared
-/// against the commitment computed by the VM.
+/// Verifies the proof together with all deferred precompile requests.
 ///
-/// # Returns
-/// Returns a tuple `(security_level, aggregated_commitment)` where:
-/// - `security_level`: The security level (in bits) of the verified proof
-/// - `aggregated_commitment`: A [`Word`] containing the final aggregated commitment to all
-///   precompile requests, computed by recomputing and recording each precompile commitment in a
-///   transcript. This value is the finalized digest of the recomputed precompile transcript.
-///
-/// # Errors
-/// Returns any error produced by [`verify`], as well as any errors resulting from precompile
-/// verification.
-#[tracing::instrument("verify_program", skip_all)]
+/// This helper recomputes the precompile commitments using the supplied
+/// [`PrecompileVerifierRegistry`], rebuilds the transcript, and verifies the STARK proof
+/// with the recomputed transcript state as part of the public inputs.
 pub fn verify_with_precompiles(
     program_info: ProgramInfo,
     stack_inputs: StackInputs,
     stack_outputs: StackOutputs,
     proof: ExecutionProof,
-    precompile_verifiers: &PrecompileVerifierRegistry,
+    registry: &PrecompileVerifierRegistry,
 ) -> Result<(u32, PrecompileTranscriptDigest), VerificationError> {
-    // get security level of the proof
     let security_level = proof.security_level();
-    let program_hash = *program_info.program_hash();
+    let (hash_fn, proof_bytes, precompile_requests) = proof.into_parts();
 
-    let (hash_fn, proof, precompile_requests) = proof.into_parts();
+    // Recompute the precompile transcript by verifying all precompile requests
+    let transcript = registry.requests_transcript(&precompile_requests)?;
```

Comment:
In this file overall there are a few changes that could be reverted as they shouldn't have an impact on verification. It's fine for now but we may be losing some context and making the diff larger.